src/types.rs

```diff
@@ -0,0 +1,131 @@
+use pyo3::{
+    prelude::*,
+    types::{PyDict, PyList},
+    PyNativeType, ToBorrowedObject, ToPyObject,
+};
+
+pub trait PyArray<'py>: Into<PyObject> + Clone + 'py {
```

Comment:
I think if you add `PyTryFrom` and `PyTypeInfo`, then you'll get access to `PyAny::downcast` and `PyAny::is_instance` without having to require them on these traits. Seems like it's probably a useful idea?